algorithms/src/sorting/impl/Kokkos_SortByKeyImpl.hpp

```diff
@@ -0,0 +1,322 @@
+      KOKKOS_LAMBDA(int i) { permute(i) = i; });
+
+  static_assert(sizeof...(MaybeComparator) <= 1);
+  if constexpr (sizeof...(MaybeComparator) == 0) {
+    Kokkos::sort(
+        exec, permute,
+        KOKKOS_LAMBDA(int i, int j) { return keys(i) < keys(j); });
+  } else {
+    auto keys_comparator =
+        std::get<0>(std::tuple<MaybeComparator...>(maybeComparator...));
+    Kokkos::sort(
+        exec, permute, KOKKOS_LAMBDA(int i, int j) {
+          return keys_comparator(keys(i), keys(j));
+        });
+  }
+
+  applyPermutation(exec, permute, keys);
+  applyPermutation(exec, permute, values);
+}
+
+// ------------------------------------------------------
+//
+// specialize cases for sorting by key without comparator
+//
+// ------------------------------------------------------
+
+#if defined(KOKKOS_ENABLE_CUDA)
+template <class KeysDataType, class... KeysProperties, class ValuesDataType,
+          class... ValuesProperties>
+void sort_by_key_device_view_without_comparator(
+    const Kokkos::Cuda& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values) {
+  sort_by_key_cudathrust(exec, keys, values);
+}
+#endif
+
+#if defined(KOKKOS_ENABLE_ROCTHRUST)
+template <class KeysDataType, class... KeysProperties, class ValuesDataType,
+          class... ValuesProperties>
+void sort_by_key_device_view_without_comparator(
+    const Kokkos::HIP& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values) {
+  sort_by_key_hipthrust(exec, keys, values);
+}
+#endif
+
+#if defined(KOKKOS_ENABLE_ONEDPL)
+template <class DataType, class... Properties>
+void sort_by_key_device_view_without_comparator(
+    const Kokkos::Experimental::SYCL& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values) {
+  if (keys.stride(0) == 1 && values.stride(0) == 1) {
+    sort_by_key_onedpl(exec, keys, values);
+  } else {
+    sort_by_key_via_binsort(exec, keys, values);
+  }
+}
+#endif
+
+// fallback case
+template <class ExecutionSpace, class KeysDataType, class... KeysProperties,
+          class ValuesDataType, class... ValuesProperties>
+std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value>
+sort_by_key_device_view_without_comparator(
+    const ExecutionSpace& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values) {
+  sort_by_key_via_sort(exec, keys, values);
+}
+
+// ---------------------------------------------------
+//
+// specialize cases for sorting by key with comparator
+//
+// ---------------------------------------------------
+
+#if defined(KOKKOS_ENABLE_CUDA)
+template <class ComparatorType, class KeysDataType, class... KeysProperties,
+          class ValuesDataType, class... ValuesProperties>
+void sort_by_key_device_view_with_comparator(
```

Comment:
I don't know if I should duplicate these or not. I followed the way it's written in `Kokkos_SortImpl.hpp`.